### PR TITLE
[JIT] ARM64 vector right-shift must dead-code eliminate if the constant is zero.

### LINKED ISSUE #89868: Incorrect codegen on Arm64
While looking into https://github.com/dotnet/runtime/issues/89338, the `op_RightShift` tests on Arm64 showed failures for a subset of scenarios.

All of the results end up as `0` even though `left` is correctly initialized:
```
17:42:31.190 Running test: _Vector128_1_r::JIT.HardwareIntrinsics.General._Vector128_1.Program.op_RightShiftByte8()
Beginning scenario: RunBasicScenario_UnsafeRead
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunBasicScenario_UnsafeRead failed:
    left: (245, 22, 238, 240, 15, 153, 34, 47, 250, 145, 133, 253, 229, 154, 2, 185)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

Beginning scenario: RunReflectionScenario_UnsafeRead
Beginning scenario: RunClsVarScenario
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunClsVarScenario failed:
    left: (112, 122, 66, 152, 94, 156, 161, 126, 15, 225, 23, 71, 143, 171, 193, 40)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

Beginning scenario: RunLclVarScenario_UnsafeRead
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunLclVarScenario_UnsafeRead failed:
    left: (245, 22, 238, 240, 15, 153, 34, 47, 250, 145, 133, 253, 229, 154, 2, 185)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

Beginning scenario: RunClassLclFldScenario
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunClassLclFldScenario failed:
    left: (43, 37, 34, 218, 97, 164, 96, 166, 121, 198, 230, 210, 76, 33, 195, 101)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

Beginning scenario: RunClassFldScenario
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunClassFldScenario failed:
    left: (138, 243, 196, 71, 173, 47, 7, 6, 216, 3, 151, 118, 183, 16, 46, 196)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

Beginning scenario: RunStructLclFldScenario
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunStructLclFldScenario failed:
    left: (83, 50, 3, 81, 81, 39, 143, 250, 230, 126, 255, 164, 121, 205, 106, 133)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

Beginning scenario: RunStructFldScenario
Vector128.op_RightShift<Byte>(Vector128<Byte>, 8): RunStructFldScenario failed:
    left: (206, 100, 99, 24, 132, 104, 161, 61, 175, 93, 168, 77, 20, 92, 227, 44)
  result: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
```

The extracted `RunStructLclFldScenario` test shows the following codegen:
```asm
; Method VectorImmBinaryOpTest:RunStructLclFldScenario():this (FullOpts)
G_M18422_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x40]!
            stp     x19, x20, [sp, #0x20]
            stp     x21, x22, [sp, #0x30]
            mov     fp, sp
            mov     x19, x0
						;; size=20 bbWeight=1 PerfScore 4.00

G_M18422_IG02:  ;; offset=0x0014
            movz    x0, #0xE4D8      // code for VectorImmBinaryOpTest+TestStruct:Create():VectorImmBinaryOpTest+TestStruct
            movk    x0, #0xFD89 LSL #16
            movk    x0, #0x7FF7 LSL #32
            ldr     x0, [x0]
            blr     x0
            stp     x0, x1, [fp, #0x10]	// [V01 loc0], [V01 loc0+0x08]
            add     x20, x19, #24
            mov     x21, x20
            ldrsb   wzr, [x21]
            add     x22, x21, #32
            mov     x0, x22
            movz    x1, #0xE400      // code for System.Runtime.InteropServices.GCHandle:AddrOfPinnedObject():long:this
            movk    x1, #0xFD26 LSL #16
            movk    x1, #0x7FF7 LSL #32
            ldr     x1, [x1]
            blr     x1
            ldr     x1, [x21, #0x10]
            add     x0, x0, x1
            sub     x0, x0, #1
            sub     x1, x1, #1
            bic     x0, x0, x1
            ldr     q16, [fp, #0x10]	// [V13 tmp10]
            str     q16, [x0]
            mov     x0, x22
            movz    x1, #0xE400      // code for System.Runtime.InteropServices.GCHandle:AddrOfPinnedObject():long:this
            movk    x1, #0xFD26 LSL #16
            movk    x1, #0x7FF7 LSL #32
            ldr     x1, [x1]
            blr     x1
            ldr     x1, [x20, #0x10]
            add     x0, x0, x1
            sub     x0, x0, #1
            sub     x1, x1, #1
            bic     x1, x0, x1
            mov     x0, x19
            ldr     q0, [fp, #0x10]	// [V13 tmp10]
            movz    x2, #0x9550
            movk    x2, #0x8020 LSL #16
            movk    x2, #458 LSL #32
            movz    x3, #0x49C0      // code for VectorImmBinaryOpTest:ValidateResult(System.Runtime.Intrinsics.Vector128`1[ubyte],ulong,System.String):this
            movk    x3, #0xFD6F LSL #16
            movk    x3, #0x7FF7 LSL #32
            ldr     x3, [x3]
            blr     x3
						;; size=176 bbWeight=1 PerfScore 45.50

G_M18422_IG03:  ;; offset=0x00C4
            ldp     x21, x22, [sp, #0x30]
            ldp     x19, x20, [sp, #0x20]
            ldp     fp, lr, [sp], #0x40
            ret     lr
						;; size=16 bbWeight=1 PerfScore 4.00
; Total bytes of code: 212
```

If we mark `DataTable.get_outArrayPtr` as `NoInline` the issue goes away.

Notably the `ShiftRight` in this scenario is a `no-op` because it is simply "shift 0" and it correctly gets optimized out in the above, so the bug seems to be elsewhere and around the value stored in the `fp, #0x10` local.

### PR DESCRIPTION
**Pull request by TIHan**
Fixes https://github.com/dotnet/runtime/issues/89868

The vector right-shift instructions do not allow a zero immediate value in the encoding. See the [docs](https://developer.arm.com/documentation/dui0801/g/A64-SIMD-Vector-Instructions/USHR--vector-).

> shift
Is the right shift amount, in the range 1 to the element width in bits, and can be one of the values shown in Usage.

### CODE REVIEW

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 1]
```diff
@@ -335,29 +335,32 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+                HWIntrinsicImmOpHelper helper(this, op, node);
 
                 for (helper.EmitBegin(); !helper.Done(); helper.EmitCaseEnd())
                 {
                     const int shiftAmount = helper.ImmValue();
 
-                    GetEmitter()->emitIns_R_R_I(ins, emitSize, targetReg, op2Reg, shiftAmount, opt);
+                    if (shiftAmount == 0)
+                    {
+                        GetEmitter()->emitIns_R_R_R(INS_mov, emitTypeSize(node), targetReg, reg, reg);
```
Comment (TIHan) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:347]: We really want to use `emitIns_Mov` but there is an issue with the ARM64 impl that will elide the `mov` even if `canSkip` is `false`.

Comment (MichalPetryka) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:350]: Why is eliding a mov an issue here?

Comment (tannergooding) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:350]: It breaks the jump table logic which expects a fixed number of bytes per switch case (on Arm64 that's 2 instructions per case, so 8-bytes)

**Comment by TIHan** [comment]
@dotnet/jit-contrib @BruceForstall this is ready.

**Review by tannergooding — APPROVED** [review]

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**